src/main/daml/Daml/Finance/Equity/Instrument.daml

```diff
@@ -44,15 +44,13 @@ template Instrument
     implements Equity.I where
       asInstrument = toInterface @Instrument.I this
       view = Equity.View ()
-      distributionImpl Equity.Distribution{id; perUnitDistribution; effectiveDate; newInstrument} key =
-        coerceContractId <$>
-          create Lifecycle.Distribution with issuer; effectiveDate; perUnitDistribution; currentInstrument = key; newInstrument; id; observers
-      stockSplitImpl Equity.StockSplit{id; factor; newInstrument; effectiveDate} key =
-        coerceContractId <$>
-          create Lifecycle.StockSplit with issuer; currentInstrument = key; factor; newInstrument; effectiveDate; id; observers
-      replacementImpl Equity.Replacement{id; offerer; perUnitPayout; effectiveDate} key =
-        coerceContractId <$>
-          create Lifecycle.Replacement with offerer; executionDate = effectiveDate; targetInstrument = key; offeredAssets = perUnitPayout; id; observers
+      declareDividendImpl Equity.DeclareDividend{id; effectiveDate; newInstrument; perUnitDistribution} = do
+        toInterfaceContractId <$> create Distribution.Event with provider = issuer; id; effectiveDate; targetInstrument = instrumentKey; newInstrument; perUnitDistribution; observers
+      declareStockSplitImpl Equity.DeclareStockSplit{id; adjustmentFactor; newInstrument; effectiveDate} =
+        -- NOTE: Doesn't handle conversion of fractional shares into cash
```

Comment:
This could be handled directly in the `Lifecyclable` rule used to create the `Effect`.

Reply:
True - but I think we need to decide if we want to tackle the topic of precision, rounding, and unit size holistically throughout the codebase.